anomaly_detectors/anomalib_lmi/base.py

```diff
@@ -36,21 +36,95 @@ class Anomalib_Base(ABC):
     
     logger = logging.getLogger('Anomalib Base')
     logger.setLevel(logging.INFO)
-    
     tiler = None
+    inference_mode = None
+    fp16 = False
+
+    if torch.cuda.is_available():
+        device = torch.device('cuda:0')
+    else:
+        logger.warning('GPU device unavailable. Use CPU instead.')
+        device = torch.device('cpu')
     
     @abstractmethod
     def __init__(self) -> None:
-        pass
-    
+        pass    
     
     @torch.inference_mode()
     def from_numpy(self, x):
         """
         convert numpy array to torch tensor
         """
         return torch.from_numpy(x).to(self.device) if isinstance(x, np.ndarray) else x
+    
+    @torch.inference_mode()
+    def warmup(self,input_hw=None):
+        '''
+        Desc: 
+            Warm up model using a np zeros array with shape matching model input size.
+        Args: 
+            input_hw(int | list, optional): a int if h equals to w, or a list of [h,w]. Need to specify this if using tiling. Otherwise, use model's built-in shape.
+        '''
+        if input_hw is None:
+            input_hw = self.model_shape
+        input_hw = to_list(input_hw)
+        zeros = np.zeros(input_hw+[3,])
+        self.predict(zeros)
+
+    @torch.inference_mode()
+    def preprocess(self, image):
+        '''
```

Comment:
I would recommend each class has its own preprocess method. So,
1. It would have easier logic, easier to maintain/upgrade.
2. remove some class attributes: such as self.inference_mode.